borsh-rs/borsh/src/de/mod.rs

```diff
@@ -1,69 +1,66 @@
+use crate::Input;
 use std::collections::{BTreeMap, HashMap, HashSet};
-use std::io::{Cursor, Error, Read};
+use std::io::{Error};
 use std::mem::size_of;
 
 mod hint;
 
-const ERROR_NOT_ALL_BYTES_READ: &str = "Not all bytes read";
-
 /// A data-structure that can be de-serialized from binary format by NBOR.
 pub trait BorshDeserialize: Sized {
-    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error>;
+    fn deserialize<I: Input>(input: &mut I) -> Result<Self, Error>;
 
     /// Deserialize this instance from a slice of bytes.
     fn try_from_slice(v: &[u8]) -> Result<Self, Error> {
-        let mut c = Cursor::new(v);
-        let result = Self::deserialize(&mut c)?;
-        if c.position() != v.len() as u64 {
-            return Err(std::io::Error::new(
-                std::io::ErrorKind::InvalidData,
-                ERROR_NOT_ALL_BYTES_READ,
-            ));
-        }
+        let mut input = vec![0; v.len()];
```

Comment:
This is not efficient, we are already given a slice, why do we need to copy all bytes from the slice into an intermediate vector?

Reply:
In that case we'd have to change an interface `fn try_from_slice(v: &[u8]) -> Result<Self, Error>` to `fn try_from_slice(v: &mut [u8]) -> Result<Self, Error>`.
If one wants to pass slice directly he can use `fn deserialize<I: Input>(input: &mut I) -> Result<Self, Error>;`.

Reply:
Why the following does not work?
```
fn try_from_slice(mut v: &[u8]) -> Result<Self, Error> {
Self::deserialize(&mut v)
}
```

Reply:
that works!